File: Csharp_class/cpp_bank_system/src/LineOfCreditAccount.cpp
#include "../include/LineOfCreditAccount.h"

// 构造函数实现
LineOfCreditAccount::LineOfCreditAccount(const std::string& name, double initialBalance, double creditLimit)
    : BankAccount(name, initialBalance, -creditLimit) {  // 将信用限额设为负的最小余额
}

// 重写月末交易处理方法
void LineOfCreditAccount::performMonthEndTransactions() {
    if (getBalance() < 0) {
        // 将余额取负值以获得正的利息费用
        double interest = -getBalance() * 0.07;  // 7%利息
        makeWithdrawal(interest, std::chrono::system_clock::now(), "按月收取利息");
    }
}

// 重写透支检查方法（体现模板方法模式和多态性）
// 这里完全重写了基类的行为：不抛出异常，而是收取费用
std::unique_ptr<Transaction> LineOfCreditAccount::checkWithdrawalLimit(bool isOverdrawn) {
    if (isOverdrawn) {
        // 返回透支费交易，而不是抛出异常
        return std::make_unique<Transaction>(-20.0, std::chrono::system_clock::now(), "申请透支费");
    } else {
        return nullptr;  // 对应C#的default
    }
}

File: Csharp_class/LineOfCreditAccount.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
/*信用帐户：
1.余额可以为负，但不能大于信用限额的绝对值。
2.如果月末余额不为 0，每个月都会产生利息。
3.将在超过信用限额的每次提款后收取费用。

*/
namespace Csharp_class
{
    internal class LineOfCreditAccount : BankAccount
    {
        public LineOfCreditAccount(string name, decimal initialBalance, decimal creditLimit) : base(name,initialBalance, -creditLimit)
        {
        }

        public override void PerformMonthEndTransactions()
        {
            if (Balance < 0)
            {
                // Negate the balance to get a positive interest charge:
                var interest = -Balance * 0.07m;
                MakeWithdrawal(interest, DateTime.Now, "按月收取利息");
            }
        }
        //当帐户透支时，该重写将返回费用交易。 如果取款未超出限额，则该方法将返回 null 交易。
        protected override Transaction? CheckWithdrawalLimit(bool isOverdrawn) =>
        isOverdrawn? new Transaction(-20, DateTime.Now, "申请透支费") : default;

    }
}
